File: apps/mobile/lib/src/features/study/data/study_api_client.dart
import 'dart:convert';

import 'package:dio/dio.dart';

import '../../../core/network/api_client.dart';
import '../../../core/network/api_envelope.dart';
import '../../../core/network/api_error.dart';
import '../domain/annotation_token.dart';
import '../domain/lookup_result.dart';
import '../domain/translation_result.dart';

class StudyApiClient {
  StudyApiClient({Dio? dio, ApiClient? apiClient})
    : _dio = dio ?? apiClient?.dio ?? ApiClient().dio;

  final Dio _dio;

  Future<LookupResult> lookup({
    required String text,
    required String sourceLang,
    required String targetLang,
    String? context,
  }) async {
    final data = await _request(
      'POST',
      '/api/v1/study/lookup',
      data: {
        'text': text,
        'sourceLang': sourceLang,
        'targetLang': targetLang,
        'context': ?context,
      },
    );
    return LookupResult.fromJson(data);
  }

  Future<TranslationResult> translateParagraph({
    required String text,
    required String sourceLang,
    required String targetLang,
  }) async {
    final data = await _request(
      'POST',
      '/api/v1/study/translate-paragraph',
      data: {
        'text': text,
        'sourceLang': sourceLang,
        'targetLang': targetLang,
      },
    );
    return TranslationResult.fromJson(data);
  }

  Future<List<AnnotationToken>> annotate({
    required String text,
    required String sourceLang,
  }) async {
    final data = await _request(
      'POST',
      '/api/v1/study/annotate',
      data: {'text': text, 'sourceLang': sourceLang},
    );
    final tokensJson = data['tokens'];
    if (tokensJson is! List) {
      throw const ApiError(
        code: 'INVALID_RESPONSE',
        message: 'Missing annotation tokens',
      );
    }
    return tokensJson
        .map(
          (token) =>
              AnnotationToken.fromJson(Map<String, dynamic>.from(token as Map)),
        )
        .toList(growable: false);
  }

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    required Map<String, Object?> data,
  }) async {
    final response = await _dio.request<Object?>(
      path,
      data: data,
      options: Options(method: method, validateStatus: (_) => true),
    );

    final envelope = ApiEnvelope.fromJson(_decode(response.data));
    if (!envelope.success) {
      throw _mapError(
        envelope.error ??
            const ApiError(code: 'UNKNOWN_ERROR', message: 'Request failed'),
      );
    }

    final envelopeData = envelope.data;
    if (envelopeData == null) {
      throw const ApiError(
        code: 'INVALID_RESPONSE',
        message: 'Missing response data',
      );
    }
    return envelopeData;
  }

  ApiError _mapError(ApiError error) {
    return switch (error.code) {
      'TRANSLATION_UNSUPPORTED_LANGUAGE_PAIR' => ApiError(
        code: 'STUDY_UNSUPPORTED_LANGUAGE_PAIR',
        message: error.message,
      ),
      'TRANSLATION_PROVIDER_UNAVAILABLE' => ApiError(
        code: 'STUDY_PROVIDER_UNAVAILABLE',
        message: error.message,
      ),
      _ => error,
    };
  }

  Object? _decode(Object? data) {
    if (data is String) {
      return jsonDecode(data);
    }
    return data;
  }
}

File: apps/web-reader/services/studyApiClient.ts
import { ofetch } from 'ofetch'

import type { ApiEnvelope } from '../types/api'
import { WebApiError } from '../types/api'
import type { WebLexeme } from '../types/localData'

type Fetcher = (url: string, options: Record<string, unknown>) => Promise<unknown>

export class WebStudyApiError extends WebApiError {}

export interface StudyApiClientOptions {
  baseUrl?: string
  fetcher?: Fetcher
}

export interface LookupRequest {
  text: string
  sourceLang: string
  targetLang: string
  context?: string
}

export type PublicLexemeSnapshot = Omit<WebLexeme, 'createdAt' | 'updatedAt'>

export interface LookupResult {
  kind: string
  lexeme: PublicLexemeSnapshot
  provider?: string | null
  providerMessage?: string | null
}

export interface TranslateParagraphRequest {
  text: string
  sourceLang: string
  targetLang: string
}

export interface TranslationResult {
  translatedText: string
  provider?: string | null
  cached?: boolean
  message?: string | null
}

export interface StudyApiClient {
  lookup(input: LookupRequest): Promise<LookupResult>
  translateParagraph(input: TranslateParagraphRequest): Promise<TranslationResult>
}

export function createStudyApiClient(options: StudyApiClientOptions = {}): StudyApiClient {
  const baseUrl = options.baseUrl ?? ''
  const fetcher = options.fetcher ?? ofetch
  return {
    lookup(input) {
      return request<LookupResult>(fetcher, baseUrl, '/api/v1/study/lookup', {
        method: 'POST',
        body: {
          text: input.text,
          sourceLang: input.sourceLang,
          targetLang: input.targetLang,
          ...(input.context ? { context: input.context } : {}),
        },
      })
    },
    translateParagraph(input) {
      return request<TranslationResult>(fetcher, baseUrl, '/api/v1/study/translate-paragraph', {
        method: 'POST',
        body: {
          text: input.text,
          sourceLang: input.sourceLang,
          targetLang: input.targetLang,
        },
      })
    },
  }
}

async function request<T>(
  fetcher: Fetcher,
  baseUrl: string,
  path: string,
  options: Record<string, unknown>,
): Promise<T> {
  const envelope = await fetcher(`${baseUrl}${path}`, options) as ApiEnvelope<T>
  if (!envelope.success) {
    const error = envelope.error ?? { code: 'UNKNOWN_ERROR', message: 'Request failed' }
    throw mapStudyError(error.code, error.message)
  }
  if (envelope.data == null) {
    throw new WebStudyApiError('WEB_INVALID_RESPONSE', 'Missing response data')
  }
  return envelope.data
}

function mapStudyError(code: string, message: string): WebStudyApiError {
  if (code === 'TRANSLATION_PROVIDER_UNAVAILABLE') {
    return new WebStudyApiError('WEB_TRANSLATION_PROVIDER_UNAVAILABLE', message)
  }
  if (code === 'TRANSLATION_UNSUPPORTED_LANGUAGE_PAIR') {
    return new WebStudyApiError('WEB_TRANSLATION_UNSUPPORTED_LANGUAGE_PAIR', message)
  }
  return new WebStudyApiError(code, message)
}
